codec-http/src/main/java/io/netty/handler/codec/http/routing/DualRouter.java

```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright 2014 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http.routing;
+
+/** Similar to Router, but the target can be both class or instances of the class. */
+public abstract class DualRouter<M, T, RouterLike extends DualRouter<M, T, RouterLike>> {
+    protected abstract RouterLike getThis();
+
+    protected abstract M CONNECT();
+    protected abstract M DELETE();
+    protected abstract M GET();
+    protected abstract M HEAD();
+    protected abstract M OPTIONS();
+    protected abstract M PATCH();
+    protected abstract M POST();
+    protected abstract M PUT();
+    protected abstract M TRACE();
+
+    //--------------------------------------------------------------------------
+
+    class ObjectRouter extends Router<M, Object, ObjectRouter> {
```

Comment:
I'll add "final". The non "static" is "on purpose", to attach the inner class instances to the outer class instances.